README.md
SHIELD [XSH] Source Code
================================

Specifications:
--------------

* Symbol: XSH
* PoW (proof of work)
* Algorithms: scrypt, x17, Lyra2rev2, myr-groestl, & blake2s
* Blocktime: 15 seconds -
* RPC port: 20103 -
* P2P port: 21103 -
* Blockreward: 
  * Block 0 to 14,000 : 8,000 coins
  * 14,000 to 28,000 : 4,000 coins
  * 28,000 to 42,000: 2,000 coins
  * 42,000 to 210,000: 1,000 coins
  * 210,000 to 378,000: 500 coins
  * 378,000 to 546,000: 250 coins
  * 546,000 to 714,000: 125 coins
  * 714,000 to 2,124,000: 62 coins
  * 2,124,000 to 4,248,000: 32 coins

Total Supply
------------

Approximately total reward cap: ~660 Million -

Binary (pre-compiled) wallets are available on all platforms at [https://ShieldCoin.github.io](https://ShieldCoin.github.io/#wallets)

Compiling Linux Wallet on Ubuntu/Debian (faster) 
----------------------

sudo rm -Rf ~/SHIELD  #(if you already have it) 

sudo apt-get -y install git && cd ~ && git clone https://github.com/ShieldCoin/SHIELD && cd SHIELD && sh go.sh


Compiling Linux Wallet on Ubuntu/Debian
----------------------

Step 1. Install the dependencies. 

```sudo apt-get update```

```sudo apt-get install libdb-dev libdb++-dev build-essential libtool autotools-dev automake pkg-config libssl1.0-dev libevent-dev bsdmainutils git libboost-all-dev libminiupnpc-dev libqt5gui5 libqt5core5a libqt5webkit5-dev libqt5dbus5 qttools5-dev qttools5-dev-tools libprotobuf-dev protobuf-compiler libqrencode-dev```

**Note**: If you are on debian, you will also need to `apt-get install libcanberra-gtk-module`.

Step 2. Clone the git repository and compile the daemon and gui wallet:

```git clone https://github.com/ShieldCoin/SHIELD && cd SHIELD && ./autogen.sh && ./configure --with-incompatible-bdb && make```

if you are using source-build libdb4.8(++)-dev you may need to use

```./configure  CPPFLAGS="-I/usr/local/BerkeleyDB.4.8/include -O2" LDFLAGS="-L/usr/local/BerkeleyDB.4.8/lib"```

**Note**: If you get a "memory exhausted" error, make a swap file. (https://www.digitalocean.com/community/tutorials/how-to-add-swap-space-on-ubuntu-16-04)

Using the wallet:
----
The gui wallet is in ./SHIELD/src/qt and the daemon in ./SHIELD/src directories.

**Note**: If you see something like 'Killed (program cc1plus)' run ```dmesg``` to see the error(s)/problems(s). This is most likely caused by running out of resources. You may need to add some RAM or add some swap space.

**Optional**:
If you want to copy the binaries for use by all users, run the following commands:

```sudo cp src/SHIELDd /usr/bin/```

```sudo cp src/qt/SHIELD-qt /usr/bin/```

Step 3. Creating a configuration file. Type ```cd ~``` to get back to the home folder and type:

```SHIELDd.exe```  (or ```./SHIELDd``` if on mac or linux)

the output from this command will tell you that you need to make a Shield.conf and will suggest some good starting values.

For Linux users, type:
 
```nano ~/.SHIELD/SHIELD.conf```
(For Windows users, see below. For mac users, the command is ```nano ~/Library/Application\ Support\SHIELD\SHIELD.conf```)
    
Paste the output from the `SHIELDd` command into the Shield.conf like this: (It is recommended to change the password to something unique.)

    rpcuser=Shieldrpcusername
    rpcpassword=85CpSuCNvDcYsdQU8w621mkQqJAimSQwCSJL5dPT9wQX
    
    
**Optional**: Add `rpcport=20103`, `port=21103`, or `algo=groestl` to the configuration file.

Add `daemon=1`. 

Your config may look something like this:

    rpcuser=Shieldrpcusername
    rpcpassword=85CpSuCNvDcYsdQU8w621mkQqJAimSQwCSJL5dPT9wQX
    rpcport=20103
    port=21103
    daemon=1
    algo=groestl

Exit the Shield.conf by pressing `ctrl + x` on your keyboard then pressing `y` and hitting enter. This should have created a Shield.conf file with what you just added. 

Type ```SHIELDd.exe``` (or ```./SHIELDd``` if on mac or linux) and your Shield daemon should start.

To check the status of how much of the blockchain has been downloaded (aka synced) type `SHIELDd.exe getinfo` (or `./SHIELDd getinfo` if on mac or linux).



To compile on Mac (OSX El Capitan, but test compiled on Mountain Lion v10.8):
------------
1. Ensure you do not have qt5 nor qt installed.

    `brew uninstall qt qt5 qt55 qt52`   
   
2. Download [qt5.4.2](https://download.qt.io/official_releases/qt/5.4/5.4.2/qt-opensource-mac-x64-clang-5.4.2.dmg)
     
3. Install qt5 into /usr/local/qt5
   
   Note: Change the installation folder from "xxx/Qt5.4.2" to "/usr/local/qt5"
   
4. Run these commands:

    `export PKG_CONFIG_PATH=/usr/local/qt5/5.4/clang_64/lib/pkgconfig`
    
    `export PATH=/usr/local/qt5/5.4/clang_64/bin:$PATH`
    
    `export QT_CFLAGS="-I/usr/local/qt5/5.4/clang_64/lib/QtWebKitWidgets.framework/Versions/5/Headers -I/usr/local/qt5/5.4/clang_64/lib/QtWebView.framework/Versions/5/Headers -I/usr/local/qt5/5.4/clang_64/lib/QtDBus.framework/Versions/5/Headers -I/usr/local/qt5/5.4/clang_64/lib/QtWidgets.framework/Versions/5/Headers -I/usr/local/qt5/5.4/clang_64/lib/QtWebKit.framework/Versions/5/Headers -I/usr/local/qt5/5.4/clang_64/lib/QtNetwork.framework/Versions/5/Headers -I/usr/local/qt5/5.4/clang_64/lib/QtGui.framework/Versions/5/Headers -I/usr/local/qt5/5.4/clang_64/lib/QtCore.framework/Versions/5/Headers -I. -I/usr/local/qt5/5.4/clang_64/mkspecs/macx-clang -F/usr/local/qt5/5.4/clang_64/lib"`
    
    `export QT_LIBS="-F/usr/local/qt5/5.4/clang_64/lib -framework QtWidgets -framework QtGui -framework QtCore -framework DiskArbitration -framework IOKit -framework OpenGL -framework AGL -framework QtNetwork -framework QtWebKit -framework QtWebKitWidgets -framework QtDBus -framework QtWebView"`
    
    
5. Install the other required items:

    `brew install protobuf boost miniupnpc openssl qrencode berkeley-db4 automake`
    
6. Download the wallet source and build:

    `git clone https://github.com/ShieldCoin/SHIELD`

    `cd SHIELD`
    
    `./autogen.sh`
    
    `./configure --with-gui=qt5`
    
    `make -j4`
    

Building the Mac installer (.dmg) file
-----
Run `make deploy`

If you are building the .dmg (by running 'mac deploy') you may need to run these commands if you get an error regarding mysql:

    brew install mysql
    
    cd /usr/local/qt5/5.4/clang_64/plugins/sqldrivers
    
    otool -L libqsqlmysql.dylib

Note: This may be pointing to an version of mysql that you do not have installed (like mysql55) - Alternatively, you may be able to remove the library from the sqldrivers folder.

    install_name_tool -change /opt/local/lib/mysql55/mysql/libmysqlclient.18.dylib /usr/local/Cellar/mysql/5.7.12/lib/libmysqlclient.20.dylib libqsqlmysql.dylib
    
Trying to build .dmg on 10.8? You will need to run this:
    
    export CFLAGS=-Qunused-arguments
    
    export CPPFLAGS=-Qunused-arguments
    
    sudo -E easy_install appscript
    

Want to use Docker?
------------

Check out the [readme](https://github.com/ShieldCoin/Shield/tree/master/contrib/docker) for more information.


Want to 'solo-mine' from the wallet?
----------
You can use the wallet to mine. You need to specify the algorithm (see below) and set the "gen" flag. For instance, in the configuration specify ```gen=1```. 


Using different algorithms (for mining)
----------

To use a specific mining algorithm use the `algo` switch in your configuration file (.conf file) or from the command line (like this `--algo=x17`) Here are the possible values:

    algo=x17
    algo=scrypt
    algo=groestl
    algo=lyra
    algo=blake

Using SHIELD on Windows
-------------

1. Download the pre-compiled software. (only from official ShieldCoin site)
2. Install
3. In windows file explorer, open c:\Users\XXX\AppData\Roaming\ShieldCoin (be sure to change XXX to your windows user)
4. Right click and create a new file SHIELD.txt
5. Edit the file to have contents above (see instructions above for options)
6. Save and close the file
7. Reame the file to ShieldCoin.conf
8. Start the SHIELD-qt program.
9. Open up SHIELD-qt console and run ```getinfo``` (or ```getmininginfo```) to verify settings.

**Note:** You must re-start the wallet after making changes to ShieldCoin.conf.
[How To build on Windows](https://bitcointalk.org/index.php?topic=149479.0)


Thanks
----
Special Thanks to: 
* VERGE Devs
* TOR Devs
* Bitcoin Devs
* Tor Devs


Linux Wallet Video Tutorial
-------
https://www.youtube.com/watch?v=WYe75b6RWes




releases.txt
# see http://stackoverflow.com/questions/12343452/how-to-publish-artifacts-in-travis-ci/33109519

To create a release, tag the commit:
   git commit -m 'some commit message'
   git tag -m 'new shiny release' 2.0.0.2 c02aa6e8ae407f3693a2fb26acbd0108d311cd71
   git push
   git push --tags

To delete a tag:
   git tag -d 2.0.0.2

To see tags: ('-n' shows comment, '-l' shows in order)
   git tag -n -l


-- Minimum OS for installation validation
- Mac OSX 10.8 Mountain Lion
- Windows 7
- Ubuntu 14.04
  - For Ubuntu 14.04, you need to run these steps before you can run SHIELD-qt:
    sudo apt-get install -y libboost-filesystem-dev libboost-program-options-dev libboost-thread-dev libqrencode-dev
    sudo add-apt-repository ppa:bitcoin/bitcoin
    sudo apt-get update
    sudo apt-get install libdb4.8++-dev


go.sh
#!/bin/bash
#// full deployement : run sh go.sh

# generating entropy make it harder to guess the randomness!.
echo "Initializing random number generator..."
random_seed=/var/run/random-seed
# Carry a random seed from start-up to start-up
# Load and then save the whole entropy pool
if [ -f $random_seed ]; then
    sudo cat $random_seed >/dev/urandom
else
    sudo touch $random_seed
fi
sudo chmod 600 $random_seed
poolfile=/proc/sys/kernel/random/poolsize
[ -r $poolfile ] && bytes=`sudo cat $poolfile` || bytes=512
sudo dd if=/dev/urandom of=$random_seed count=1 bs=$bytes

#Also, add the following lines in an appropriate script which is run during the$

# Carry a random seed from shut-down to start-up
# Save the whole entropy pool
echo "Saving random seed..."
random_seed=/var/run/random-seed
sudo touch $random_seed
sudo chmod 600 $random_seed
poolfile=/proc/sys/kernel/random/poolsize
[ -r $poolfile ] && bytes=`sudo cat $poolfile` || bytes=512
sudo dd if=/dev/urandom of=$random_seed count=1 bs=$bytes

# Create a swap file

cd ~
if [ -e /swapfile1 ]; then
echo "Swapfile already present"
else
sudo dd if=/dev/zero of=/swapfile1 bs=1024 count=524288
sudo mkswap /swapfile1
sudo chown root:root /swapfile1
sudo chmod 0600 /swapfile1
sudo swapon /swapfile1
fi

# Install dependency

sudo apt-get -y install software-properties-common

sudo add-apt-repository -y ppa:bitcoin/bitcoin

sudo apt-get update

sudo apt-get -y install libcanberra-gtk-module

# Dont need to check if bd is already installed, will override or pass by
#results=$(find /usr/ -name libdb_cxx.so)
#if [ -z $results ]; then
sudo apt-get -y install libdb4.8-dev libdb4.8++-dev
#else
#grep DB_VERSION_STRING $(find /usr/ -name db.h)
#echo "BerkeleyDb will not be installed its already there...."
#fi

sudo apt-get -y install git build-essential libtool autotools-dev autoconf automake pkg-config libssl-dev libevent-dev bsdmainutils git libprotobuf-dev protobuf-compiler libqrencode-dev

sudo apt-get -y install libqt5gui5 libqt5core5a libqt5webkit5-dev libqt5dbus5 qttools5-dev qttools5-dev-tools

sudo apt-get -y install libminiupnpc-dev

# Keep current version of libboost if already present
results=$(find /usr/ -name libboost_chrono.so)
if [ -z $results ]; then
sudo apt-get -y install libboost-all-dev
else
red=`tput setaf 1`
green=`tput setaf 2`
reset=`tput sgr0`
echo "${red}Libboost will not be installed its already there....${reset}"
grep --include=*.hpp -r '/usr/' -e "define BOOST_LIB_VERSION"
fi

sudo apt-get -y install --no-install-recommends gnome-panel

sudo apt-get -y install lynx

sudo apt-get -y install unzip

cd ~

#// Compile Berkeley if 4.8 is not there
if [ -e /usr/lib/libdb_cxx-4.8.so ]
then
echo "BerkeleyDb already present...$(grep --include *.h -r '/usr/' -e 'DB_VERSION_STRING')" 
else
wget http://download.oracle.com/berkeley-db/db-4.8.30.NC.tar.gz 
tar -xzvf db-4.8.30.NC.tar.gz 
rm db-4.8.30.NC.tar.gz
cd db-4.8.30.NC/build_unix 
../dist/configure --enable-cxx 
make 
sudo make install 
sudo ln -s /usr/local/BerkeleyDB.4.8/lib/libdb-4.8.so /usr/lib/libdb-4.8.so
sudo ln -s /usr/local/BerkeleyDB.4.8/lib/libdb_cxx-4.8.so /usr/lib/libdb_cxx-4.8.so
cd ~
sudo rm -Rf db-4.8.30.NC
#sudo ldconfig
fi

#// Check if libboost is present

results=$(find /usr/ -name libboost_chrono.so)
if [ -z $results ]; then
sudo rm download
     wget https://sourceforge.net/projects/boost/files/boost/1.63.0/boost_1_63_0.zip/download 
     unzip -o download
     cd boost_1_63_0
	sh bootstrap.sh
	sudo ./b2 install
	cd ~
	sudo rm download 
	sudo rm -Rf boost_1_63_0
	#sudo ln -s $(dirname "$(find /usr/ -name libboost_chrono.so)")/lib*.so /usr/lib
	sudo ldconfig
        #sudo rm /usr/lib/libboost_chrono.so
else
     echo "Libboost found..." 
     grep --include=*.hpp -r '/usr/' -e "define BOOST_LIB_VERSION"
fi

#// Clone files from repo, Permissions and make

git clone https://github.com/ShieldCoin/shield
cd SHIELD
sudo sh autogen.sh
chmod 777 ~/SHIELD/share/genbuild.sh
chmod 777 ~/SHIELD/src/leveldb/build_detect_platform



grep --include=*.hpp -r '/usr/' -e "define BOOST_LIB_VERSION"

sudo rm wrd01.txt
sudo rm wrd00.txt
sudo rm words
find /usr/ -name libboost_chrono.so > words
split -dl 1 --additional-suffix=.txt words wrd



if [ -e wrd01.txt ]; then
echo 0. $(cat wrd00.txt)
echo 1. $(cat wrd01.txt)
echo 2. $(cat wrd02.txt)
echo 3. $(cat wrd03.txt)
echo -n "Choose libboost library to use(0-3)?"
read answer
else
echo "There is only 1 libboost library present. We choose for you 0"
answer=0
fi

echo "You have choosen $answer"

if [ -d /usr/local/BerkeleyDB.4.8/include ]
then
./configure CPPFLAGS="-I/usr/local/BerkeleyDB.4.8/include -O2" LDFLAGS="-L/usr/local/BerkeleyDB.4.8/lib" --with-gui=qt5 --with-boost-libdir=$(dirname "$(cat wrd0$answer.txt)")
echo "Using Berkeley Generic..."
else
./configure --with-gui=qt5 --with-boost-libdir=$(dirname "$(cat wrd0$answer.txt)")
echo "Using default system Berkeley..."
fi

make -j$(nproc)

if [ -e ~/SHIELD/src/qt/SHIELD-qt ]; then
#sudo apt-get -y install pulseaudio
#sudo apt-get -y install portaudio19-dev
# synthetic voice 
#cd ~
#wget https://sourceforge.net/projects/espeak/files/espeak/espeak-1.48/espeak-1.48.04-source.zip/download
#unzip -o download
#cd espeak-1.48.04-source/src
#cp portaudio19.h portaudio.h
#make
#cd ~
for i in 528 1000 1600 2000 3000
do
pactl load-module module-sine frequency=$i > /dev/null
sleep 0.1
done
sleep 1
pactl unload-module module-sine
pactl upload-sample complet test2
openssl rand -hex 4096 | padsp tee /dev/audio > /dev/null
sleep 1.3
pactl play-sample test2
#espeak mission,complete
sudo strip ~/SHIELD/src/SHIELDd
sudo strip ~/SHIELD/src/qt/SHIELD-qt
sudo make install
else
echo "Compile fail not SHIELD-qt present"
fi

cd ~

#// Create the config file with random user and password

mkdir ~/.SHIELD
cp ~/.SHIELD/SHIELD.conf ~/.SHIELD/SHIELD.bak
echo "rpcuser="$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 26 ; echo '') '\n'"rpcpassword="$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 26 ; echo '') '\n'"rpcport=20103" '\n'"port=21103" '\n'"daemon=1" '\n'"listen=1" '\n'"server=1" '\n'"addnode=103.18.40.125:21103" '\n'"addnode=104.131.144.82:21103" '\n'"addnode=138.197.68.130:21103" '\n'"addnode=144.76.167.66:21103" '\n'"addnode=152.186.36.86:21103" '\n'"addnode=159.203.121.202:21103" '\n'"addnode=172.104.157.38:21103" '\n'"addnode=192.99.7.127:21103" '\n'"addnode=219.89.84.46:21103" '\n'"addnode=45.32.129.168:21103" '\n'"addnode=45.55.59.206:21103" '\n'"addnode=46.4.64.68:21103" '\n'"addnode=51.15.46.1:21103" '\n'"addnode=52.9.109.214:21103" '\n'"addnode=66.55.64.183:21103" '\n'"addnode=67.167.207.164:21103" '\n'"addnode=78.46.190.152:21103"> ~/.SHIELD/SHIELD.conf

#// Extract http link, download blockchain and install it.

echo -n "Do you wish to download the complete SHIELD Blockchain (y/n)?"
read answer
if echo "$answer" | grep -iq "^y" ;then
    sudo rm SHIELD-Blockchain*.zip
    until [ -e SHIELD*.zip ]
    do
    sleep 1
    echo "wget" $(lynx --dump --listonly https://ShieldCoin.de | grep -o "https:*.*zip") > link.sh
    sleep 1
    sh link.sh
    done
    
    #checksum
    sudo rm blockchain
    wget https://www.ShieldCoin.github.io/checksums/blockchain
    md5sum SHIELD-Blockchain*.zip > md5
    checksum="$(grep $(cat md5) blockchain)"
    if [ -z "$checksum" ];
    then
    echo "Warning: MD5 is not matching"
    else
    echo "MD5 is matching...Success"
    fi
    
    unzip -o SHIELD-Blockchain*.zip -d ~/.SHIELD
    sudo rm SHIELD-Blockchain*.zip
else
 echo "Blockchain will not be installed sync may be long"   
fi

# Create Icon on Desktop and in menu

sudo cp ~/SHIELD/src/qt/res/icons/SHIELD.png /usr/share/icons/
echo '#!/usr/bin/env xdg-open''\n'"[Desktop Entry]"'\n'"Version=1.0"'\n'"Type=Application"'\n'"Terminal=false"'\n'"Icon[en]=/usr/share/icons/SHIELD.png"'\n'"Name[en]=SHIELD"'\n'"Exec=SHIELD-qt"'\n'"Name=SHIELD"'\n'"Icon=/usr/share/icons/SHIELD.png"'\n'"Categories=Network;Internet;" > ~/Desktop/SHIELD.desktop
sudo chmod +x ~/Desktop/SHIELD.desktop
sudo cp ~/Desktop/SHIELD.desktop /usr/share/applications/SHIELD.desktop
sudo chmod +x /usr/share/applications/SHIELD.desktop

# Erase all SHIELD compilation directory , cleaning

cd ~
#sudo rm -Rf ~/SHIELD

#// Start SHIELD

SHIELD-qt
cp ~/.SHIELD/wallet.dat ~/SHIELDwallet.bak
